cpp/dolfinx/fem/interpolate.h

```diff
@@ -107,8 +147,495 @@ void interpolate_from_any(Function<T>& u, const Function<T>& v)
+      const auto yp = x_t(i, 1);
+      const auto zp = x_t(i, 2);
+
+      for (const auto& p : candidates[i])
+      {
+        pointsToSend[p].push_back(xp);
+        pointsToSend[p].push_back(yp);
+        pointsToSend[p].push_back(zp);
+      }
+    }
+
+    std::vector<std::int32_t> nPointsToSend(nProcs);
+    std::transform(pointsToSend.cbegin(), pointsToSend.cend(),
+                   nPointsToSend.begin(),
+                   [](const auto& el) { return el.size(); });
+
+    std::vector<std::int32_t> allPointsToSend(nProcs * nProcs);
+    MPI_Allgather(nPointsToSend.data(), nProcs, MPI_INT32_T,
+                  allPointsToSend.data(), nProcs, MPI_INT32_T, MPI_COMM_WORLD);
+
+    std::size_t nPointsToReceive = 0;
+    for (int i = 0; i < nProcs; ++i)
+    {
+      nPointsToReceive += allPointsToSend[mpi_rank + i * nProcs];
+    }
+    std::vector<std::int32_t> sendingOffsets(nProcs, 0);
+    for (int i = 0; i < nProcs; ++i)
+    {
+      for (int j = 0; j < mpi_rank; ++j)
+      {
+        sendingOffsets[i] += allPointsToSend[nProcs * j + i];
+      }
+    }
+
+    xt::xtensor<double, 2> pointsToReceive(
+        xt::shape(
+            {nPointsToReceive / 3, static_cast<decltype(nPointsToReceive)>(3)}),
+        0);
+    MPI_Win window;
+    MPI_Win_create(pointsToReceive.data(), sizeof(double) * nPointsToReceive,
+                   sizeof(double), MPI_INFO_NULL, MPI_COMM_WORLD, &window);
+    MPI_Win_fence(0, window);
+
+    for (int i = 0; i < nProcs; ++i)
+    {
+      MPI_Put(pointsToSend[i].data(), pointsToSend[i].size(), MPI_DOUBLE, i,
+              sendingOffsets[i], pointsToSend[i].size(), MPI_DOUBLE, window);
+    }
+
+    MPI_Win_fence(0, window);
+    MPI_Win_free(&window);
+
+    // Each process will now check at which points it can evaluate
+    // the interpolating function, and note that down in evaluationCells
+    std::vector<std::int32_t> evaluationCells(pointsToReceive.shape(0), -1);
+
+    const auto connectivity = v.function_space()->mesh()->geometry().dofmap();
+
+    // This BBT is useful for fast lookup of which cell contains a given point
+    dolfinx::geometry::BoundingBoxTree bbt(*v.function_space()->mesh(), tdim,
+                                           0.0001);
+
+    const auto xv = v.function_space()->mesh()->geometry().x();
+
+    if (tdim == 3)
+    {
+      // For each point at which the source function needs to be evaluated
+      for (decltype(pointsToReceive.shape(0)) i = 0;
+           i < pointsToReceive.shape(0); ++i)
+      {
+        // Get its coordinates
+        const double xp = pointsToReceive(i, 0);
+        const double yp = pointsToReceive(i, 1);
+        const double zp = pointsToReceive(i, 2);
+
+        // For each cell that might contain that point
+        for (const auto& j :
+             dolfinx::geometry::compute_collisions(bbt, {xp, yp, zp}))
+        {
+          // Get the vertexes that belong to that cell
+          const auto vtx = connectivity.links(j);
```

Comment:
Is there as specific reason for you are not using: https://github.com/FEniCS/dolfinx/blob/d316ef77c58883f09d0e68486aa1caa24a10f6ec/cpp/dolfinx/geometry/utils.h#L83-L94

Especially as the code below seems to only work for tetrahedrons.

Reply:
I didn't know of this function, I'll check it out. Thanks!

Reply:
And yes, the code is only meant to work for simplicial meshes in 2 and 3 dimensions. Does FEniCS-X support non-simplicial meshes too? I never thought of it as I only ever used triangles and tetras :_D

Reply:
Yes, FEniCSx supports, quads and hexes (and soon prisms as @chrisrichardson  is working on it)

Reply:
Ok, I looked into that function and it turns out I had reinvented the wheel once again :D
Thanks for pointing that out, I switched to using it of course.